app/server/utils/ussd/kenya_ussd_state_machine.py

```diff
@@ -86,7 +86,6 @@ def change_preferred_language_to_en(self, user_input):
 
     def change_preferred_language_to(self, language):
         self.user.preferred_language = language
-        db.session.commit()
```

Comment:
if i have this line in here, the tests give me errors saying:
```
sqlalchemy.exc.IntegrityError: (psycopg2.IntegrityError) null value in column "session_id" violates not-null constraint
```

I'm guessing because it's trying to create the factories..? if i don't do this line, the user change still sticks for the test. does that mean it also stuck for good? i'm still a bit confused when we need to commit given there's that magic code somewhere that makes committing not necessary

Reply:
I am running into this very error with my tests.

Reply:
@enjeyw any idea what the way out here would be ?

Reply:
>  i'm still a bit confused when we need to commit given there's that magic code somewhere that makes committing not necessary

Oh that magic code is only for requests going the the REST API - these changes still need a commit. 

> does that mean it also stuck for good?

Nope, just for the duration of that test module. The following code in conftest tears down all ur changes:
```

@pytest.fixture(scope='module')
def init_database():
    # Create the database and the database table

    with current_app.app_context():
        db.create_all()  # todo- use manage.py

    yield db  # this is where the testing happens!

    with current_app.app_context():
        db.session.close_all()  # DO NOT DELETE THIS LINE. We need to close sessions before dropping tables.
        db.drop_all()
```

Reply:
at it's most basic, this is failing because the column session_id needs a value. @PhilipWafula have u tried just inserting a random string to make it happy?

Reply:
Yeah I fed it a value same error.

Reply:
errrgg that's annoying. ummmm i'm at a loss unfortunately.

Reply:
@PhilipWafula per https://stackoverflow.com/a/19664562, we need to do UssdSessionFactory.build instead of UssdSessionFactory()
i've fixed mine in this PR

Reply:
build gives it an auto-incremented id so it doesn't error. but this begs the question how do we test if a db change actually stuck around vs we forgot to db.session.commit?

Reply:
oh wait i mis-read. we're supposed to use `create` and not `build`. `build` does NOT save it to the database. i think the problem is we didn't set session_id as the primary key. will fix right now

Reply:
OKAY IGNORE EVERYTHING I SAID. it's an early morning, my brain hasn't fully booted. session_id is not a primary key, it's passed in from the USSD API but needs to be not null and needs to be unique, so i've defined it as a sequence on the factory (see `UssdSessionFactory`). there were other fields that had to be not null as well so i also faked those. NOTE @PhilipWafula: i've just defined `ussd_menu_id` as 1. for later more complicated tests we may need that to be tied to actual ussd menus, but it'll do for now.
what i don't get though is according to the docs, `UssdSessionFactory()` is the same as `UssdSessionFactory.create()`. in that case, it should be saving to the db already. but then how come the null constraint errors didn't happen until we had code that manually did a `db.session.create()` in the test?

Reply:
just tried it with UssdSessionFactory.create... also doesn't error until our explicit db commit. whyyyy? their documentation says "returns a saved instance", but it clearly did not save properly if it had violated null constraint errors?? or is this some sqlalchemy or something else black magic where initial saves are okay to violate constraints...

Reply:
Ohhhhh I think i can actually answer this one!

basically the state that u store in an SQLAlchemy session isn't actually validated against db constraints until you commit to db. it's not so much black magic of SQLA telling u it's ok, so much as it is SQLAbeing totally ignorant of what the db has done till then. The reasoning being it's basically impossible for SQLA to know for certain whether something violates constraints without talking to the db ("sure this key was unique when i last check the db state, but how do i know another agent hasn't added a row with the same key since then").

Reply:
ooo makes sense!